db/couchdb/design_docs/items.json

```diff
@@ -25,6 +25,9 @@
     },
     "byShelvesAndListing": {
       "map": "(doc)->\n  if doc.shelves?\n    for shelf in doc.shelves\n      emit [shelf, doc.listing], null"
+    },
+    "byListingWithoutShelf": {
+      "map": "(doc)->\n  if ! doc.shelves?\n    emit [doc.owner,doc.listing], null"
```

Comment:
From what I've seen, this would work because the IDs are unique even between collection (or whatever it is called in couchdb) but it's still a "dirty hack" in a way, as the name can't let you predict what it would do and the answer would still look like the expected result.
Using this modified request would mean changing all the `build_keys_from_shelf`, or the `byShelf`/`byShelves` code, further obfuscating the intent.

I understand the need to reduce code duplication, but in this case I'd suggest to propagate the `noShelf` parameter to `getAuthorizedItems.byUser` then to `getByAccessLevel`.

What do you think?

Reply:
It could be made more explicit that this is a pseudo shelf id based on a user id, example:
```js
const defaultUserShelf = { _id: `u:${userId}:default`, owner: userId }
```
and that wouldn't change anything in `build_keys_from_shelf` as far as I can tell. What I like about this solution is that it underlines that it's a shelf concern. We could even push the logic to replace the new `no-shelf` parameter by a user-id based pseudo-shelf-id that would then carry that role through the whole stack, from the client API call to the database view, and could have other yet-unknown applications.

But it does introduce a new concept (pseudo-shelf-ids), which would require to update validation, and might create some weirdness here and there, so if the solution you suggest to propagate the `noShelf` parameter works and is less weird, we can go with that :)

On another subject but still related to those CouchDB views, beware that you need to check for the case where `doc.shelves= []`, which should be considered to be a 'no shelf' case
```suggestion
    "byShelvesAndListing": {
      "map": "(doc)->\n  if doc.shelves?.length > 0 \n    for shelf in doc.shelves\n      emit [shelf, doc.listing], null"
    },
    "byListingWithoutShelf": {
      "map": "(doc)->\n  if (not doc.shelves?) or doc.shelves.length is 0\n    emit [doc.owner,doc.listing], null"
```

Reply:
I made the change for `doc.shelves.length is 0` and propagated the `withoutShelf` argument (I renamed it for consistency with the view)